[1LP][RFR] Storage Manager [Block and Object] WT

digitronik opened the pull request:
Purpose or Intent
=================
- added combine navigation destination for object and block storage manager.
- added simple navigation test for block and object storage manager.


{{pytest: cfme/tests/storage/test_manager.py -v --use-provider rhos7-ga}}

dajoRH commented:
Lint report for commit cd21cbec3007d82d0a66b86c3bc8ef8312b2fc3e:

`cfme/storage/manager.py`:
- :bangbang: Line 193:45: **E127** *continuation line over-indented for visual indent*
- :bangbang: Line 193:69: **E231** *missing whitespace after ','*
- :bangbang: Line 215:69: **E231** *missing whitespace after ','*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

quarckster commented on cfme/storage/manager.py:
```diff
@@ -0,0 +1,239 @@
+                                    FlashMessages
+                                    )
+from widgetastic.widget import View, NoSuchElementException, Text
+from cfme.exceptions import ItemNotFound
```
**REQUIRED**
Please follow style guide for imports.

quarckster commented on cfme/storage/manager.py:
```diff
@@ -0,0 +1,239 @@
+    """The all Storage Manager or Provider page"""
+    @property
+    def is_displayed(self):
+        """Is this page currently being displayed"""
```
Here and below a docstring for `is_displayed` property not needed.

quarckster commented on cfme/tests/storage/test_manager.py:
```diff
@@ -0,0 +1,48 @@
+@pytest.mark.uncollectif(lambda provider: not provider.one_of(OpenStackProvider))
+def test_block_manager_navigation(openstack_provider, appliance):
+    """Simple test for navigation destinations"""
+    block_name = openstack_provider.name + ' Cinder Manager'
```
**REQUIRED**
Please use `format`

dajoRH commented:
Lint report for commit 80f662c0121307c398dffc20e1d6739d597728f7:

`cfme/storage/manager.py`:
- :bangbang: Line 15:5: **E123** *closing bracket does not match indentation of opening bracket's line*
- :bangbang: Line 20:5: **E123** *closing bracket does not match indentation of opening bracket's line*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

digitronik commented:
@quarckster added required changes.

digitronik commented:
@quarckster collection added. Due to the impact of the collection, i removed the excessive block and object manages classes. The required attributes added in respective collection classes only.

quarckster commented on cfme/storage/manager.py:
```diff
@@ -0,0 +1,233 @@
+            self.breadcrumb.active_location == "'Storage Manager' Policy Assignment")
+
+
+class BlockManagerCollection(NavigatableMixin):
```
**REQUIRED**
Please read today's last email from @psav http://cfme-tests.readthedocs.io/guides/collections.html

digitronik commented:
@quarckster  improved with a new collection policy, remove navMixin and added BaseCollection and BaseEntity with other changes.

quarckster commented on cfme/tests/storage/test_manager.py:
```diff
@@ -0,0 +1,51 @@
+def test_block_manager_navigation(openstack_provider, appliance):
+    """Simple test for navigation destinations"""
+
+    collection = BlockManagerCollection(appliance=appliance)
```
**REQURIED**
This should be in a fixture.

quarckster commented on cfme/tests/storage/test_manager.py:
```diff
@@ -0,0 +1,51 @@
+
+
+@pytest.mark.uncollectif(lambda provider: not provider.one_of(OpenStackProvider))
+def test_object_manager_navigation(openstack_provider, appliance):
```
**REQUIRED**
These tests are identical, so it's better to have only one with the name let's say ` test_managers_navigation` and parametrized fixture.

digitronik replied:
@quarckster  added.

izapolsk commented on cfme/storage/manager.py:
```diff
@@ -0,0 +1,233 @@
+    configuration = Dropdown('Configuration')
+    policy = Dropdown('Policy')
+    monitoring = Dropdown('Monitoring')
+    download = Button(title='Download summary in PDF format')
```
QUESTION: are you sure it always has such title ? 
Besides that, are you sure this is Button not Dropdown ?

digitronik replied:
yes, it's not a dropdown, a simple button with pdf icon, having title "Download summary in PDF format"

izapolsk commented on cfme/storage/manager.py:
```diff
@@ -0,0 +1,233 @@
+
+    @property
+    def in_manager(self):
+        nav = self.context['object'].nav.pick(self.context['object'].appliance.version)
```
REQUIRED: this is good idea to move navigation path to class but  I don't like this name - "nav". It doesn't explain what it contains even more it sounds similar to navmazing object/property. I'd advice to call it nav_path or navigation_path for example.

izapolsk commented on cfme/storage/manager.py:
```diff
@@ -0,0 +1,233 @@
+    select_value = BootstrapSelect('tag_add')
+    save_button = Button('Save')
+    reset_button = Button('Reset')
+    cancel = Button('Cancel')
```
Please either add _button ending everywhere or don't use it at all

izapolsk commented on cfme/storage/manager.py:
```diff
@@ -0,0 +1,233 @@
+        return Manager(self, name, provider)
+
+
+class Manager(BaseEntity):
```
REQUIRED: Please use full name. class name, etc should always be self explaining. it makes code easily readable.

izapolsk commented on cfme/storage/manager.py:
```diff
@@ -0,0 +1,233 @@
+    Args:
+        name: Name of the object manager.
+        provider: provider
+        appliance: appliance
```
REQUIRED: there is no such argument now

izapolsk commented on cfme/storage/manager.py:
```diff
@@ -0,0 +1,233 @@
+        self.provider = provider
+        self.nav = self.collection.nav
+        self.type = self.collection.type
+        self.name_vari = VersionPick({Version.lowest(): 'Storage Provider',
```
REQUIRED: please give readable name to variable

izapolsk commented on cfme/storage/manager.py:
```diff
@@ -0,0 +1,233 @@
+                                               handle_alert=not cancel)
+
+        if not cancel:
+            msg = "Refresh Provider initiated for 1 {} from the CFME Database"\
```
REQUIRED: please get rid of back slash

izapolsk commented on cfme/tests/storage/test_manager.py:
```diff
@@ -0,0 +1,44 @@
+    view = navigate_to(manager, 'Details')
+    assert view.is_displayed
+
+    manager.refresh()
```
Q: what is it for ?

digitronik replied:
@izapolsk, Its completely new page so added simple navigation test. while navigating added refresh method to capture its flash [Idea came when I added one BZ for the same refresh from 'All' Page]. this test going to remove as I start automating tests for this page.

digitronik commented:
@izapolsk changes committed. Please review as you free.

izapolsk commented on cfme/storage/manager.py:
```diff
@@ -0,0 +1,233 @@
+    entities = View.nested(ManagerDetailsEntities)
+
+
+class ManagerTagsView(ManagerView):
```
@psav , @rlbabyuk, tags/policies views are the same everywhere. Do we have any such shared views to re-use here ?

digitronik replied:
@izapolsk   Yup you are right, I referred @rlbabyuk  PR about TagPageView and WidgetasticTaggable. removing it.

izapolsk commented on cfme/storage/manager.py:
```diff
@@ -0,0 +1,233 @@
+    """The policies page for Storage Manager or Provider"""
+    breadcrumb = BreadCrumb()
+    policies = BootstrapTreeview("protectbox")
+    save_button = Button("Save")
```
REQUIRED: as I already wrote before please either use _button everywhere or remove it everywhere.

izapolsk commented on cfme/storage/manager.py:
```diff
@@ -0,0 +1,233 @@
+
+    def __init__(self, appliance):
+        self.appliance = appliance
+        self.nav = VersionPick({
```
REQUIRED: as I already wrote before, please use another name

izapolsk commented on cfme/storage/manager.py:
```diff
@@ -0,0 +1,233 @@
+        self.appliance = self.collection.appliance
+        self.name = name
+        self.provider = provider
+        self.nav = self.collection.nav
```
^^

digitronik replied:
@izapolsk, I have to take nav [navigation_path] attribute from the collection. it's helping me to suppress "super method" with two different classes.

izapolsk replied:
I already wrote you about nav name. it is confusing and doesn't explain what it contains.

izapolsk commented on cfme/storage/manager.py:
```diff
@@ -0,0 +1,233 @@
+        return BaseManager(self, name, provider)
+
+
+class BaseManager(BaseEntity):
```
REQUIRED: can you please use some self explaining name here ? for example StorageManager, etc.

If you run into BaseManager class somewhere in another file, you won't be able to get what is that.
So, class names or public properties should have self explaining names.

dajoRH commented:
Lint report for commit c1d659f6d158177393c374f28b084347759ab966:

`cfme/storage/manager.py`:
- :bangbang: Line 79:65: **E126** *continuation line over-indented for hanging indent*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

digitronik commented:
@izapolsk , changes
nav > naviagation_path
Manager > StorageManager, as well as all Views, name also changed.
added, WidgetasticTaggeble method to avoid tag view.

Ready for Review

izapolsk commented:
super, now test name is much more readable
